File: postgres-indexed-implementation/loader.go
package main

import (
	"encoding/csv"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

var (
	policyHeader = []string{"policy_id", "tenant_id", "department_id", "default_action", "enabled", "created_at", "updated_at"}
	ruleHeader   = []string{"rule_id", "policy_id", "priority", "rule_type", "action", "enabled", "regex_pattern", "source_ip_group_key", "start_time_utc", "end_time_utc", "created_at", "updated_at"}
	groupHeader  = []string{"group_id", "tenant_id", "department_id", "group_key", "name", "created_at", "updated_at"}
	memberHeader = []string{"group_id", "source_ipv4", "created_at", "updated_at"}
)

func LoadEngine(dataDirectory string) (*Engine, error) {
	return loadEngine(csvMasterSource{directory: dataDirectory})
}

type masterSource interface {
	dataVersion() (string, error)
	readPolicies(func([]string) error) error
	readRules(func([]string) error) error
	readGroups(func([]string) error) error
	readMembers(func([]string) error) error
}

type csvMasterSource struct{ directory string }

func (source csvMasterSource) dataVersion() (string, error) {
	return readDatasetVersion(source.directory)
}

func readDatasetVersion(dataDirectory string) (string, error) {
	manifest, err := os.ReadFile(filepath.Join(dataDirectory, "manifest.json"))
	if err != nil {
		return "", fmt.Errorf("read dataset manifest: %w", err)
	}
	var metadata struct {
		DatasetVersion string `json:"dataset_version"`
	}
	if err := json.Unmarshal(manifest, &metadata); err != nil || metadata.DatasetVersion == "" {
		return "", errors.New("invalid dataset manifest")
	}
	return metadata.DatasetVersion, nil
}

func (source csvMasterSource) readPolicies(consume func([]string) error) error {
	return readCSVDirectory(filepath.Join(source.directory, "policies"), policyHeader, consume)
}

func (source csvMasterSource) readRules(consume func([]string) error) error {
	return readCSVDirectory(filepath.Join(source.directory, "firewall_rules"), ruleHeader, consume)
}

func (source csvMasterSource) readGroups(consume func([]string) error) error {
	return readCSVDirectory(filepath.Join(source.directory, "source_ip_groups"), groupHeader, consume)
}

func (source csvMasterSource) readMembers(consume func([]string) error) error {
	return readCSVDirectory(filepath.Join(source.directory, "source_ip_group_members"), memberHeader, consume)
}

func loadEngine(source masterSource) (*Engine, error) {
	dataVersion, err := source.dataVersion()
	if err != nil {
		return nil, err
	}
	engine := &Engine{tenants: make(map[string]*tenantData), DataVersion: dataVersion}
	policiesByID := make(map[string]*Policy)
	policyScopes := make(map[string]struct{})
	if err := source.readPolicies(func(row []string) error {
		id, tenantID, departmentID := row[0], row[1], row[2]
		enabled, enabledErr := strconv.ParseBool(row[4])
		if id == "" || tenantID == "" || !actionValid(row[3]) || enabledErr != nil {
			return fmt.Errorf("invalid or unsupported policy %q", id)
		}
		if _, exists := policiesByID[id]; exists {
			return fmt.Errorf("duplicate policy_id %q", id)
		}
		scope := tenantID + "\x00" + departmentID
		if _, exists := policyScopes[scope]; exists {
			return fmt.Errorf("duplicate policy scope for %s/%s", tenantID, departmentID)
		}
		policyScopes[scope] = struct{}{}
		policy := newPolicy(id, row[3])
		policiesByID[id] = policy
		if !enabled {
			return nil
		}
		tenant := engine.tenant(tenantID)
		if departmentID == "" {
			if tenant.policy != nil {
				return fmt.Errorf("duplicate tenant policy for %q", tenantID)
			}
			tenant.policy = policy
		} else {
			if tenant.departments[departmentID] != nil {
				return fmt.Errorf("duplicate department policy for %s/%s", tenantID, departmentID)
			}
			tenant.departments[departmentID] = policy
		}
		return nil
	}); err != nil {
		return nil, err
	}
	if err := source.readRules(func(row []string) error {
		id := row[0]
		policy := policiesByID[row[1]]
		priority64, priorityErr := strconv.ParseInt(row[2], 10, 32)
		enabled, enabledErr := strconv.ParseBool(row[5])
		if id == "" || policy == nil || priorityErr != nil || priority64 < 1 || enabledErr != nil || !actionValid(row[4]) {
			return fmt.Errorf("invalid or unsupported rule %q", id)
		}
		priority := int(priority64)
		value := candidate{priority: priority, ruleID: id, action: row[4]}
		policy.builder.priorities = append(policy.builder.priorities, int32(priority))
		switch row[3] {
		case "SOURCE_IPV4_REGEX":
			if !enabled {
				return validateRegexRule(id, row[6])
			}
			return appendRegexRule(policy, 0, row[6], value)
		case "URL_PATH_REGEX":
			if !enabled {
				return validateRegexRule(id, row[6])
			}
			return appendRegexRule(policy, 1, row[6], value)
		case "REFERER_REGEX":
			if !enabled {
				return validateRegexRule(id, row[6])
			}
			return appendRegexRule(policy, 2, row[6], value)
		case "USER_AGENT_REGEX":
			if !enabled {
				return validateRegexRule(id, row[6])
			}
			return appendRegexRule(policy, 3, row[6], value)
		case "SOURCE_IPV4_GROUP":
			if row[7] == "" {
				return fmt.Errorf("rule %q: source_ip_group_key is required", id)
			}
			if enabled {
				policy.builder.groups = append(policy.builder.groups, rawGroupRule{key: row[7], candidate: value})
			}
		case "ACCESS_TIME_RANGE":
			start, err := timeOfDaySeconds(row[8])
			if err != nil {
				return fmt.Errorf("rule %q: invalid start_time_utc: %w", id, err)
			}
			end, err := timeOfDaySeconds(row[9])
			if err != nil {
				return fmt.Errorf("rule %q: invalid end_time_utc: %w", id, err)
			}
			if start == end {
				return fmt.Errorf("rule %q: start_time_utc and end_time_utc must differ", id)
			}
			if enabled {
				policy.builder.times = append(policy.builder.times, timeRule{start: start, end: end, candidate: value})
			}
		default:
			return fmt.Errorf("rule %q: unknown rule_type %q", id, row[3])
		}
		return nil
	}); err != nil {
		return nil, err
	}
	groupsByID := make(map[string]*groupDefinition)
	if err := source.readGroups(func(row []string) error {
		id, tenantID, departmentID, key := row[0], row[1], row[2], row[3]
		if id == "" || tenantID == "" || key == "" {
			return fmt.Errorf("invalid source IP group %q", id)
		}
		if _, exists := groupsByID[id]; exists {
			return fmt.Errorf("duplicate group_id %q", id)
		}
		tenant := engine.tenant(tenantID)
		var set *groupSet
		if departmentID == "" {
			if tenant.tenantGroups == nil {
				tenant.tenantGroups = newGroupSet()
			}
			set = tenant.tenantGroups
		} else {
			set = tenant.departmentGroups[departmentID]
			if set == nil {
				set = newGroupSet()
				tenant.departmentGroups[departmentID] = set
			}
		}
		if _, exists := set.definedKeys[key]; exists {
			return fmt.Errorf("duplicate group_key %q in %s/%s", key, tenantID, departmentID)
		}
		set.definedKeys[key] = struct{}{}
		groupsByID[id] = &groupDefinition{key: key, set: set, members: make(map[string]struct{})}
		return nil
	}); err != nil {
		return nil, err
	}
	if err := source.readMembers(func(row []string) error {
		definition := groupsByID[row[0]]
		address := row[1]
		if definition == nil || address == "" {
			return fmt.Errorf("member references unknown group %q", row[0])
		}
		if _, exists := definition.members[address]; exists {
			return fmt.Errorf("duplicate member %q in group %q", address, row[0])
		}
		definition.members[address] = struct{}{}
		definition.set.keysByAddress[address] = append(definition.set.keysByAddress[address], definition.key)
		return nil
	}); err != nil {
		return nil, err
	}
	for _, policy := range policiesByID {
		if err := finalizePolicy(policy); err != nil {
			return nil, fmt.Errorf("policy %q: %w", policy.id, err)
		}
	}
	return engine, nil
}

func newGroupSet() *groupSet {
	return &groupSet{keysByAddress: make(map[string][]string), definedKeys: make(map[string]struct{})}
}

func appendRegexRule(policy *Policy, field int, pattern string, value candidate) error {
	if pattern == "" {
		return fmt.Errorf("rule %q: regex_pattern is required", value.ruleID)
	}
	policy.builder.regex[field] = append(policy.builder.regex[field], rawRegexRule{pattern: pattern, candidate: value})
	return nil
}

func validateRegexRule(ruleID, pattern string) error {
	if pattern == "" {
		return fmt.Errorf("rule %q: regex_pattern is required", ruleID)
	}
	return nil
}

func finalizePolicy(policy *Policy) error {
	builder := policy.builder
	sort.Slice(builder.priorities, func(i, j int) bool { return builder.priorities[i] < builder.priorities[j] })
	if len(builder.priorities) > 0 {
		policy.minimum = int(builder.priorities[0])
	}
	for index := 1; index < len(builder.priorities); index++ {
		if builder.priorities[index] == builder.priorities[index-1] {
			return fmt.Errorf("duplicate priority %d", builder.priorities[index])
		}
	}
	for index := range builder.regex {
		built, err := buildRegexIndex(builder.regex[index])
		if err != nil {
			return err
		}
		policy.regex[index] = built
	}
	if len(builder.groups) != 0 {
		sort.Slice(builder.groups, func(i, j int) bool {
			if builder.groups[i].key == builder.groups[j].key {
				return builder.groups[i].candidate.priority < builder.groups[j].candidate.priority
			}
			return builder.groups[i].key < builder.groups[j].key
		})
		policy.groupRules = make(map[string]candidate)
		for _, value := range builder.groups {
			if _, exists := policy.groupRules[value.key]; !exists {
				policy.groupRules[value.key] = value.candidate
			}
		}
	}
	policy.times = buildTimeIndex(builder.times)
	policy.builder = nil
	return nil
}

func timeOfDaySeconds(value string) (int, error) {
	parsed, err := time.Parse("15:04:05", value)
	if err != nil {
		return 0, err
	}
	return parsed.Hour()*3600 + parsed.Minute()*60 + parsed.Second(), nil
}

func csvFiles(directory string) ([]string, error) {
	entries, err := os.ReadDir(directory)
	if err != nil {
		return nil, err
	}
	files := make([]string, 0, len(entries))
	for _, entry := range entries {
		if !entry.IsDir() && strings.EqualFold(filepath.Ext(entry.Name()), ".csv") {
			files = append(files, filepath.Join(directory, entry.Name()))
		}
	}
	sort.Strings(files)
	if len(files) == 0 {
		return nil, fmt.Errorf("no CSV files in %s", directory)
	}
	return files, nil
}

func readCSVDirectory(directory string, expectedHeader []string, consume func([]string) error) error {
	files, err := csvFiles(directory)
	if err != nil {
		return err
	}
	for _, path := range files {
		stream, err := os.Open(path)
		if err != nil {
			return err
		}
		reader := csv.NewReader(stream)
		reader.FieldsPerRecord = len(expectedHeader)
		header, readErr := reader.Read()
		if readErr == nil && !equalStrings(header, expectedHeader) {
			readErr = fmt.Errorf("invalid CSV header %v", header)
		}
		for readErr == nil {
			row, rowErr := reader.Read()
			if errors.Is(rowErr, io.EOF) {
				break
			}
			if rowErr != nil {
				readErr = rowErr
				break
			}
			if err := consume(row); err != nil {
				readErr = err
				break
			}
		}
		closeErr := stream.Close()
		if readErr != nil {
			return fmt.Errorf("%s: %w", path, readErr)
		}
		if closeErr != nil {
			return closeErr
		}
	}
	return nil
}

func equalStrings(left, right []string) bool {
	if len(left) != len(right) {
		return false
	}
	for index := range left {
		if left[index] != right[index] {
			return false
		}
	}
	return true
}

File: postgres-indexed-java8-implementation/src/main/java/firewall/Loader.java
package firewall;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.sql.SQLException;
import java.time.LocalTime;
import java.time.format.DateTimeParseException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Builds an immutable {@link Engine} from any {@link MasterSource}. Mirrors the
 * priority-indexed loader: same-condition rules keep only the minimum priority
 * (others can never be First Match), exact indexes use sorted arrays, time
 * indexes switch to the dense point index at >=64 distinct ranges.
 */
final class Loader {
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private Loader() {
    }

    static Engine load(MasterSource source) throws IOException, SQLException, InterruptedException {
        String dataVersion = source.dataVersion();
        Map<String, TenantData> tenants = new HashMap<>();
        Map<String, Policy> policiesById = new HashMap<>();
        Set<String> policyScopes = new HashSet<>();

        source.readPolicies(row -> {
            String id = row[0];
            String tenantId = row[1];
            String departmentId = row[2];
            String action = row[3];
            boolean enabled = parseBoolean(row[4]);
            if (id.isEmpty() || tenantId.isEmpty() || !actionValid(action) || row[4].isEmpty()) {
                throw new IOException("invalid or unsupported policy " + id);
            }
            if (policiesById.containsKey(id)) {
                throw new IOException("duplicate policy_id " + id);
            }
            String scope = tenantId + "\u0000" + departmentId;
            if (!policyScopes.add(scope)) {
                throw new IOException("duplicate policy scope for " + tenantId + "/" + departmentId);
            }
            Policy policy = new Policy(id, action);
            policiesById.put(id, policy);
            if (!enabled) {
                return;
            }
            TenantData tenant = tenantOf(tenants, tenantId);
            if (departmentId.isEmpty()) {
                if (tenant.policy != null) {
                    throw new IOException("duplicate tenant policy for " + tenantId);
                }
                tenant.policy = policy;
            } else {
                if (tenant.departments.containsKey(departmentId)) {
                    throw new IOException("duplicate department policy for " + tenantId + "/" + departmentId);
                }
                tenant.departments.put(departmentId, policy);
            }
        });

        source.readRules(row -> {
            String id = row[0];
            Policy policy = policiesById.get(row[1]);
            int priority;
            try {
                priority = Integer.parseInt(row[2]);
            } catch (NumberFormatException e) {
                throw new IOException("invalid or unsupported rule " + id);
            }
            boolean enabled = parseBoolean(row[5]);
            if (id.isEmpty() || policy == null || priority < 1 || row[5].isEmpty() || !actionValid(row[4])) {
                throw new IOException("invalid or unsupported rule " + id);
            }
            Policy.Builder builder = policy.builder;
            builder.priorities.add(priority);
            Candidate candidate = new Candidate(priority, id, row[4]);
            String ruleType = row[3];
            if ("SOURCE_IPV4_REGEX".equals(ruleType)) {
                addRegex(builder, Policy.SOURCE_IPV4, row[6], candidate, id);
            } else if ("URL_PATH_REGEX".equals(ruleType)) {
                addRegex(builder, Policy.URL_PATH, row[6], candidate, id);
            } else if ("REFERER_REGEX".equals(ruleType)) {
                addRegex(builder, Policy.REFERER, row[6], candidate, id);
            } else if ("USER_AGENT_REGEX".equals(ruleType)) {
                addRegex(builder, Policy.USER_AGENT, row[6], candidate, id);
            } else if ("SOURCE_IPV4_GROUP".equals(ruleType)) {
                if (row[7].isEmpty()) throw new IOException("rule " + id + ": source_ip_group_key is required");
                if (enabled) builder.groups.add(new Policy.GroupRuleRaw(row[7], candidate));
            } else if ("ACCESS_TIME_RANGE".equals(ruleType)) {
                int start = timeOfDaySeconds(row[8]);
                int end = timeOfDaySeconds(row[9]);
                if (start == end) throw new IOException("rule " + id + ": start_time_utc and end_time_utc must differ");
                if (enabled) builder.times.add(new TimeIndex.TimeRule(start, end, candidate));
            } else {
                throw new IOException("rule " + id + ": unknown rule_type " + row[3]);
            }
        });

        Map<String, GroupDefinition> groupsById = new HashMap<>();
        source.readGroups(row -> {
            String id = row[0];
            String tenantId = row[1];
            String departmentId = row[2];
            String key = row[3];
            if (id.isEmpty() || tenantId.isEmpty() || key.isEmpty()) {
                throw new IOException("invalid source IP group " + id);
            }
            if (groupsById.containsKey(id)) {
                throw new IOException("duplicate group_id " + id);
            }
            TenantData tenant = tenantOf(tenants, tenantId);
            GroupSet set;
            if (departmentId.isEmpty()) {
                if (tenant.tenantGroups == null) {
                    tenant.tenantGroups = new GroupSet();
                }
                set = tenant.tenantGroups;
            } else {
                set = tenant.departmentGroups.get(departmentId);
                if (set == null) {
                    set = new GroupSet();
                    tenant.departmentGroups.put(departmentId, set);
                }
            }
            if (!set.definedKeys.add(key)) {
                throw new IOException("duplicate group_key " + key + " in " + tenantId + "/" + departmentId);
            }
            groupsById.put(id, new GroupDefinition(key, set));
        });

        source.readMembers(row -> {
            GroupDefinition definition = groupsById.get(row[0]);
            String address = row[1];
            if (definition == null || address.isEmpty()) {
                throw new IOException("member references unknown group " + row[0]);
            }
            if (!definition.members.add(address)) {
                throw new IOException("duplicate member " + address + " in group " + row[0]);
            }
            definition.set.keysByAddress
                    .computeIfAbsent(address, k -> new ArrayList<String>())
                    .add(definition.key);
        });

        for (Policy policy : policiesById.values()) {
            finalizePolicy(policy);
        }
        return new Engine(tenants, dataVersion);
    }

    private static void addRegex(Policy.Builder builder, int field, String pattern, Candidate candidate, String ruleId) throws IOException {
        if (pattern.isEmpty()) {
            throw new IOException("rule " + ruleId + ": regex_pattern is required");
        }
        builder.regex[field].add(new Policy.RegexRuleRaw(pattern, candidate));
    }

    private static TenantData tenantOf(Map<String, TenantData> tenants, String tenantId) {
        return tenants.computeIfAbsent(tenantId, k -> new TenantData());
    }

    private static void finalizePolicy(Policy policy) {
        Policy.Builder builder = policy.builder;
        if (builder == null) {
            return;
        }
        if (builder.priorities.isEmpty()) {
            policy.minimum = Integer.MAX_VALUE;
        } else {
            List<Integer> sorted = new ArrayList<Integer>(builder.priorities);
            Collections.sort(sorted);
            policy.minimum = sorted.get(0);
            for (int i = 1; i < sorted.size(); i++) {
                if (sorted.get(i).equals(sorted.get(i - 1))) {
                    throw new RuntimeException("duplicate priority " + sorted.get(i));
                }
            }
        }
        for (int index = 0; index < builder.regex.length; index++) {
            RegexIndex.Builder regexBuilder = new RegexIndex.Builder();
            for (Policy.RegexRuleRaw raw : builder.regex[index]) {
                regexBuilder.add(raw.pattern, raw.candidate);
            }
            policy.regex[index] = regexBuilder.build();
        }
        if (!builder.groups.isEmpty()) {
            builder.groups.sort(new java.util.Comparator<Policy.GroupRuleRaw>() {
                @Override
                public int compare(Policy.GroupRuleRaw a, Policy.GroupRuleRaw b) {
                    if (!a.key.equals(b.key)) return a.key.compareTo(b.key);
                    return Integer.compare(a.candidate.priority(), b.candidate.priority());
                }
            });
            Map<String, Candidate> groupRules = new HashMap<String, Candidate>();
            for (Policy.GroupRuleRaw group : builder.groups) {
                groupRules.putIfAbsent(group.key, group.candidate);
            }
            policy.groupRules = groupRules;
        }
        TimeIndex.Builder timeBuilder = new TimeIndex.Builder();
        for (TimeIndex.TimeRule rule : builder.times) {
            timeBuilder.add(rule.start, rule.end, rule.candidate);
        }
        policy.times = timeBuilder.build();
        policy.builder = null;
    }

    private static final class GroupDefinition {
        final String key;
        final GroupSet set;
        final Set<String> members = new HashSet<String>();

        GroupDefinition(String key, GroupSet set) {
            this.key = key;
            this.set = set;
        }
    }

    static boolean actionValid(String value) {
        return "ALLOW".equals(value) || "DENY".equals(value);
    }

    static boolean parseBoolean(String value) {
        return "true".equalsIgnoreCase(value);
    }

    static int timeOfDaySeconds(String value) throws IOException {
        try {
            LocalTime parsed = LocalTime.parse(value);
            return parsed.toSecondOfDay();
        } catch (DateTimeParseException e) {
            throw new IOException("invalid time " + value, e);
        }
    }

    static String readDatasetVersion(Path dataDirectory) throws IOException {
        byte[] manifest = Files.readAllBytes(dataDirectory.resolve("manifest.json"));
        String version = MAPPER.readTree(manifest).path("dataset_version").asText("");
        if (version.isEmpty()) {
            throw new IOException("invalid dataset manifest");
        }
        return version;
    }
}
